# Exploring Hacker News Posts

*Hacker News* is a site started by the startup incubator **[Y Combinator](https://news.ycombinator.com/)**, where user-submitted stories (known as "posts") are voted and commented upon and it is extremely popular in technology and startup circles.

In this project, we will analyze the posts made by users at the website about questions asked (`Ask HN`), responses on the thread (in terms of comments) and details about projects, products (`Show HN`) or just general interesting topics. We will also compare below two types of posts in the dataset:
- Do `Ask HN` or `Show HN` receive more comments on average?
- Do posts created at a certain time receive more comments on average?

The dataset is extracted from [here](https://www.kaggle.com/hacker-news/hacker-news-posts) and contains *20,000* rows and for reference check the row below from the dataset:

| id | title | url | num_points | num_comments | author | created_at |
| -- | -- | -- | -- | -- | -- | -- |
| 12224879 | Interactive Dynamic Video | http://www.interactivedynamicvideo.com/ | 386 | 52 | ne0phyte | 8/4/2016 11:52 |

In [1]:
# import libraries
from csv import reader

# reading dataset into list of lists:
opened_file = open('hacker_news.csv')
read_file = reader(opened_file)
hn = list(read_file)
headers = hn[0]
hn = hn[1:]

# displaying header and first five rows
print(headers)
print('\n')
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


In [2]:
# creating empty lists
ask_posts = []
show_posts = []
other_posts = []

# iterating through each row to append the
# empty lists fulfilling criteria
for row in hn:
    title = row[1]
    title_lower = title.lower()
    if (title_lower.startswith('ask hn')):
        ask_posts.append(row)
    elif (title_lower.startswith('show hn')):
        show_posts.append(row)
    else:
        other_posts.append(row)

# printing lengths of all posts
print('Length of Ask Posts:', len(ask_posts))
print('Length of Show Posts:', len(show_posts))
print('Length of Other Posts:', len(other_posts))

Length of Ask Posts: 1744
Length of Show Posts: 1162
Length of Other Posts: 17194


In [3]:
# examining first five rows of `ask_posts` & `show_posts`
print(ask_posts[:5])
print('\n')
print(headers)
print('\n')
print(show_posts[:5])

[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']]


['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


[['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], ['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46'], ['11590768', 'Show HN: Shanhu

In [4]:
# calculating total number comments in 'ask_posts'
total_ask_comments = 0
for row in ask_posts:
    number_comments = row[4]
    ask_int = int(number_comments)
    total_ask_comments += ask_int

print('Total Number of ask_comments:', total_ask_comments)

Total Number of ask_comments: 24483


In [5]:
# calculating avg. number of 'ask comments'
avg_ask_comments = total_ask_comments / len(ask_posts)
avg_ask_comments

14.038417431192661

In [6]:
# calculating total number of comments in 'show_posts'
total_show_comments = 0
for row in show_posts:
    number_comments = row[4]
    show_int = int(number_comments)
    total_show_comments += show_int

print('Total Number of show_comments:', total_show_comments)

Total Number of show_comments: 11988


In [7]:
# calculating avg. number of 'show_comments'
avg_show_comments = total_show_comments / len(show_posts)
avg_show_comments

10.31669535283993

Above results show that on average comments on `ask_posts` are more than `show_posts` and this depicts that users are more engaging in these posts

In [27]:
# calculating the amount of ask_posts created in each hours of day
# along with number of comments received

import datetime as dt

result_list = []
for row in ask_posts:
    created_at = row[6]
    numb_comments = int(row[4])
    result_list.append([created_at, numb_comments])

In [29]:
# checking first 5 elements of `result_list`
result_list[:5]

[['8/16/2016 9:55', 6],
 ['11/22/2015 13:43', 29],
 ['5/2/2016 10:14', 1],
 ['8/2/2016 14:20', 3],
 ['10/15/2015 16:38', 17]]

In [30]:
# creating empty dictionaries
counts_by_hour = {}
comments_by_hour = {}
date_format = "%m/%d/%Y %H:%M"
for row in result_list:
    hour_from_date = row[0]
    comment_num = row[1]
    datetime_object = dt.datetime.strptime(hour_from_date, date_format)
    hour = datetime_object.strftime("%H")
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = comment_num
    else:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += comment_num

In [31]:
comments_by_hour

{'00': 447,
 '01': 683,
 '02': 1381,
 '03': 421,
 '04': 337,
 '05': 464,
 '06': 397,
 '07': 267,
 '08': 492,
 '09': 251,
 '10': 793,
 '11': 641,
 '12': 687,
 '13': 1253,
 '14': 1416,
 '15': 4477,
 '16': 1814,
 '17': 1146,
 '18': 1439,
 '19': 1188,
 '20': 1722,
 '21': 1745,
 '22': 479,
 '23': 543}

In [32]:
counts_by_hour

{'00': 55,
 '01': 60,
 '02': 58,
 '03': 54,
 '04': 47,
 '05': 46,
 '06': 44,
 '07': 34,
 '08': 48,
 '09': 45,
 '10': 59,
 '11': 58,
 '12': 73,
 '13': 85,
 '14': 107,
 '15': 116,
 '16': 108,
 '17': 100,
 '18': 109,
 '19': 110,
 '20': 80,
 '21': 109,
 '22': 71,
 '23': 68}

In [104]:
# calculating avg. number of comments per post during each hour of the day

avg_by_hour = []
for hr in comments_by_hour:
    avg_by_hour.append([hr, comments_by_hour[hr] / counts_by_hour[hr]])
    
avg_by_hour

[['18', 13.20183486238532],
 ['15', 38.5948275862069],
 ['07', 7.852941176470588],
 ['00', 8.127272727272727],
 ['01', 11.383333333333333],
 ['13', 14.741176470588234],
 ['11', 11.051724137931034],
 ['23', 7.985294117647059],
 ['22', 6.746478873239437],
 ['06', 9.022727272727273],
 ['16', 16.796296296296298],
 ['05', 10.08695652173913],
 ['04', 7.170212765957447],
 ['19', 10.8],
 ['21', 16.009174311926607],
 ['14', 13.233644859813085],
 ['20', 21.525],
 ['12', 9.41095890410959],
 ['09', 5.5777777777777775],
 ['17', 11.46],
 ['03', 7.796296296296297],
 ['10', 13.440677966101696],
 ['08', 10.25],
 ['02', 23.810344827586206]]

In [105]:
# swapping the elements of `avg_by_hour` list of lists
# for sorting the avg no. of comments per post

swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])

print(swap_avg_by_hour)

[[13.20183486238532, '18'], [38.5948275862069, '15'], [7.852941176470588, '07'], [8.127272727272727, '00'], [11.383333333333333, '01'], [14.741176470588234, '13'], [11.051724137931034, '11'], [7.985294117647059, '23'], [6.746478873239437, '22'], [9.022727272727273, '06'], [16.796296296296298, '16'], [10.08695652173913, '05'], [7.170212765957447, '04'], [10.8, '19'], [16.009174311926607, '21'], [13.233644859813085, '14'], [21.525, '20'], [9.41095890410959, '12'], [5.5777777777777775, '09'], [11.46, '17'], [7.796296296296297, '03'], [13.440677966101696, '10'], [10.25, '08'], [23.810344827586206, '02']]


In [106]:
# sorting the swap_avg_by_hour in descending order
# to get the highest value in list

sorted_swap = sorted(swap_avg_by_hour, reverse=True)

print(sorted_swap)

[[38.5948275862069, '15'], [23.810344827586206, '02'], [21.525, '20'], [16.796296296296298, '16'], [16.009174311926607, '21'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [13.20183486238532, '18'], [11.46, '17'], [11.383333333333333, '01'], [11.051724137931034, '11'], [10.8, '19'], [10.25, '08'], [10.08695652173913, '05'], [9.41095890410959, '12'], [9.022727272727273, '06'], [8.127272727272727, '00'], [7.985294117647059, '23'], [7.852941176470588, '07'], [7.796296296296297, '03'], [7.170212765957447, '04'], [6.746478873239437, '22'], [5.5777777777777775, '09']]


In [107]:
# printing the string
print('Top 5 Hours for `Ask HN` Comments')

# iterating through `sorted_swap` to print text using str.format()
for avg, hr in sorted_swap[:5]:
    dt_object = dt.datetime.strptime(hr, '%H')
    format_hour = dt_object.strftime('%H:%M')
    print('{}: {:.2f} average comments per post'.format(format_hour, avg))

Top 5 Hours for `Ask HN` Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

### Conclusion
In this project, we analyzed `ask posts` and `show posts` to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as `ask post` and created between 07:00 and 09:00.

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, `ask posts` received more comments on average and `ask posts` created between 07:00 and 09:00 received the most comments on average.